feat(ai): add OAuth for MCP clients + refactor to new package

aayush-kapoor opened the pull request:
## Background

Refactor mcp into it's own separate package + add OAuth for MCP clients

## Summary

Created new package `@ai-sdk/mcp`, added OAuth provider

## Manual Verification

Manual tests added, e2e example added to verify oauth works in `example/mcp/mcp-with-auth/client.ts`

## Checklist

- [x] Tests have been added / updated (for bug fixes / features)
- [ ] Documentation has been added / updated (for bug fixes / features)
- [x] A _patch_ changeset for relevant packages has been added (for bug fixes / features - run `pnpm changeset` in the project root)
- [x] Formatting issues have been fixed (run `pnpm prettier-fix` in the project root)

## Future Work

- Continue refactoring, add docs
- log warning or throw error if wrong transport protocol is set for an mcp server

## Related Issues

towards #8717, #9337 and #6717

aayush-kapoor committed:
stdio transport removed

aayush-kapoor committed:
cs

gr2m commented:
I'd have to do more research myself. I wonder why it's `ai/mcp-stdio` and not its own package? Maybe that would make sense as a first step, to remove it from `ai` core?

aayush-kapoor committed:
Merge branch 'main' into aayush/mcp-update

aayush-kapoor committed:
client oauth provider added

aayush-kapoor committed:
adding authProvider to transport config

aayush-kapoor committed:
example to iterate on

aayush-kapoor committed:
adding proper types and schema

aayush-kapoor committed:
Merge branch 'main' into aayush/mcp-update

aayush-kapoor committed:
build discovery URLs (for oauth 2.0 and oidc) to get auth server metadata

aayush-kapoor committed:
start auth flow with pkce generation

aayush-kapoor committed:
Merge branch 'main' into aayush/mcp-update

aayush-kapoor committed:
exchange auth flow added

aayush-kapoor committed:
refresh auth token flow added

aayush-kapoor committed:
register-client flow + ci fix

aayush-kapoor committed:
ci fix

aayush-kapoor committed:
add final auth function

aayush-kapoor committed:
correct auth function used

aayush-kapoor committed:
example updated to use new auth

aayush-kapoor committed:
pretty

aayush-kapoor committed:
mcp client updated

aayush-kapoor committed:
Merge branch 'main' into aayush/mcp-update

aayush-kapoor committed:
Merge branch 'main' into aayush/mcp-update

aayush-kapoor committed:
working example

aayush-kapoor committed:
http transport method added

aayush-kapoor committed:
Merge branch 'main' into aayush/mcp-update

aayush-kapoor committed:
proper imports

aayush-kapoor committed:
failing test for http transport

aayush-kapoor committed:
fix for sse resumption

aayush-kapoor committed:
working example

aayush-kapoor committed:
working example pt2

aayush-kapoor committed:
pretty

aayush-kapoor committed:
neater example

aayush-kapoor committed:
Merge branch 'main' into aayush/mcp-update

aayush-kapoor committed:
Merge branch 'main' into aayush/mcp-update

aayush-kapoor committed:
example refined

aayush-kapoor committed:
updated example - works with most mcp servers

aayush-kapoor committed:
refactor into new package

aayush-kapoor committed:
pretty

aayush-kapoor committed:
rename

aayush-kapoor committed:
changeset updated

aayush-kapoor committed:
Merge branch 'main' into aayush/mcp-update

aayush-kapoor committed:
remove pkce dependency

aayush-kapoor committed:
moving pkce to prod dep

aayush-kapoor committed:
added mcp dependency to `ai`

aayush-kapoor committed:
legacy imports still work

aayush-kapoor committed:
Merge remote-tracking branch 'origin/main' into aayush/mcp-update

aayush-kapoor committed:
pretty

aayush-kapoor committed:
build fix

aayush-kapoor committed:
deprecated fucntion ref

aayush-kapoor committed:
documentation added

aayush-kapoor committed:
pretty

aayush-kapoor committed:
Merge branch 'main' into aayush/mcp-update

aayush-kapoor committed:
agent fix

aayush-kapoor committed:
Merge branch 'main' into aayush/mcp-update

aayush-kapoor committed:
agent fix

aayush-kapoor committed:
Merge branch 'main' into aayush/mcp-update

aayush-kapoor committed:
docs fixes

aayush-kapoor committed:
bug fix

aayush-kapoor committed:
pretty

aayush-kapoor committed:
Merge branch 'main' into aayush/mcp-update

aayush-kapoor committed:
updated example

aayush-kapoor committed:
Merge branch 'aayush/mcp-update' of https://github.com/vercel/ai into aayush/mcp-update

aayush-kapoor committed:
updated zapier example for mcp

aayush-kapoor committed:
Merge branch 'main' into aayush/mcp-update

aayush-kapoor committed:
removing mcp export

aayush-kapoor committed:
working e2e exmaple

aayush-kapoor committed:
Merge branch 'main' into aayush/mcp-update

aayush-kapoor committed:
pretty + fx

aayush-kapoor committed:
Merge branch 'main' into aayush/mcp-update

aayush-kapoor committed:
agent fix

aayush-kapoor committed:
authFlow fix

aayush-kapoor committed:
Merge branch 'main' into aayush/mcp-update

gr2m reviewed (approved):
Looks great!

aayush-kapoor committed:
throw error on using wrong transport

aayush-kapoor committed:
prettier + test fix

aayush-kapoor commented on packages/mcp/src/tool/mcp-http-transport.ts:
```diff
@@ -0,0 +1,404 @@
+            }
+          } catch (error) {
+            this.onerror?.(error);
+            throw error;
```
it's needed to throw an error for users to know if they are using the wrong transport

aayush-kapoor committed:
Merge branch 'main' into aayush/mcp-update

aayush-kapoor committed:
fx

aayush-kapoor commented on examples/next-openai/app/mcp/chat/route.ts:
```diff
@@ -1,11 +1,7 @@
+import { experimental_createMCPClient } from '@ai-sdk/mcp';
 
 export async function POST(req: Request) {
   const url = new URL('http://localhost:3000/mcp/server');
```
i tested and this is working as expected if you navigate to `localhost:3000/mcp`

the server, i believe, only responds to post requests

aayush-kapoor committed:
Merge branch 'main' into aayush/mcp-update

aayush-kapoor committed:
Merge branch 'main' into aayush/mcp-update

aayush-kapoor committed:
Merge branch 'main' into aayush/mcp-update

aayush-kapoor committed:
Merge branch 'main' into aayush/mcp-update

cosmicb0y commented on examples/next-openai/app/api/mcp-with-auth/route.ts:
```diff
@@ -0,0 +1,248 @@
+        );
+      }
+    });
+  }
```
Hello, I’m not sure how this would open the authorization URL in the client’s browser — wouldn’t this just open the page on the server side?

aayush-kapoor committed:
Merge branch 'main' into aayush/mcp-update

aayush-kapoor committed:
updated example to open auth window on client side

aayush-kapoor committed:
Merge branch 'main' into aayush/mcp-update

aayush-kapoor committed:
pretty

aayush-kapoor committed:
Merge branch 'aayush/mcp-update' of https://github.com/vercel/ai into aayush/mcp-update

gr2m committed:
Merge branch 'main' into aayush/mcp-update

gr2m committed:
docs(changeset): Add OAuth for MCP clients and refactor import path

This change adds OAuth support for MCP clients and refactors the import path for Experimental_StdioMCPTransport.

gr2m committed:
Refactor import paths for MCP client and transport

Updated import paths for MCP client and transport.

gr2m reviewed (approved):
Great work!

gr2m committed:
remove remaining artifacts of mcp from `ai` package
